### PR TITLE
fix(windows): get states passing on windows

### PR DESCRIPTION
**Pull request by noelmcloughlin**
### PR progress checklist (to be filled in by reviewers)

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Describe the changes you're proposing

This PR includes changes necessary to get states passing on Windows.  

Various issues include:
- rendering error: `golang.arch`
- HTTP 404 error
- user/group error
- path errors
- salt zip handling does not like extra k/v in dict 

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### Additional context

manually tested

```
          ID: golang-package-archive-install-file-directory
    Function: file.directory
        Name: C:\\golang\\/go1.16.2.windows-amd64
      Result: True
     Comment: The directory C:\\golang\\/go1.16.2.windows-amd64 is in the correct state
     Started: 01:14:27.982567
    Duration: 4.962 ms
     Changes:
----------
          ID: golang-package-archive-install-file-directory
    Function: file.directory
        Name: C:\\golang\\packages
      Result: True
     Comment: The directory C:\\golang\\packages is in the correct state
     Started: 01:14:27.987529
    Duration: 6.0 ms
     Changes:
----------
          ID: golang-package-archive-install-archive-extracted
    Function: archive.extracted
        Name: C:\\golang\\/go1.16.2.windows-amd64
      Result: True
     Comment: All files in archive are already present
     Started: 01:14:27.994533
    Duration: 5113.292 ms
     Changes:
----------
          ID: golang-cmd-clean-cmd-run-go-clean-github.com/golang/example/hello
    Function: cmd.run
        Name: go clean -i github.com/golang/example/hello...
      Result: True
     Comment: onlyif condition is false
     Started: 01:14:33.108824
    Duration: 5040.255 ms
     Changes:
----------
          ID: golang-cmd-clean-cmd-run-go-clean-github.com/golang/example/outyet
    Function: cmd.run
        Name: go clean -i github.com/golang/example/outyet...
      Result: True
     Comment: onlyif condition is false
     Started: 01:14:38.149079
    Duration: 13.0 ms
     Changes:
----------
          ID: golang-cmd-goget-cmd-run-go-get-github.com/golang/example/hello
    Function: cmd.run
        Name: go get github.com/golang/example/hello
      Result: True
     Comment: onlyif condition is false
     Started: 01:14:38.162079
    Duration: 12.034 ms
     Changes:
----------
          ID: golang-cmd-goget-cmd-run-go-get-github.com/golang/example/outyet
    Function: cmd.run
        Name: go get github.com/golang/example/outyet
      Result: True
     Comment: onlyif condition is false
     Started: 01:14:38.174113
    Duration: 12.967 ms
     Changes:

Summary for local
------------
Succeeded: 7
Failed:    0
------------
Total states run:     7
Total run time:  10.203 s
```

### CODE REVIEW

**Comment by myii** [comment]
@noelmcloughlin The `default` suite is failing for all instances with:

```sls
                 ID: golang-cmd-goget-cmd-run-go-get-github.com/golang/example/hello
           Function: cmd.run
               Name: go get github.com/golang/example/hello
             Result: False
            Comment: Command "go get github.com/golang/example/hello" run
            Started: 08:10:54.253845
           Duration: 421.856 ms
            Changes:   
              ----------
              pid:
                  224
              retcode:
                  1
              stderr:
                  go: downloading github.com/golang/example v0.0.0-20210113200257-bcf50bfd7dcd
                  go get: github.com/golang/example@none updating to
                  	github.com/golang/example@v0.0.0-20210113200257-bcf50bfd7dcd: parsing go.mod:
                  	module declares its path as: golang.org/x/example
                  	        but was required as: github.com/golang/example
              stdout:
       ----------
                 ID: golang-cmd-goget-cmd-run-go-get-github.com/golang/example/outyet
           Function: cmd.run
               Name: go get github.com/golang/example/outyet
             Result: False
            Comment: Command "go get github.com/golang/example/outyet" run
            Started: 08:10:54.676904
           Duration: 157.091 ms
            Changes:   
              ----------
              pid:
                  239
              retcode:
                  1
              stderr:
                  go get: github.com/golang/example@none updating to
                  	github.com/golang/example@v0.0.0-20210113200257-bcf50bfd7dcd: parsing go.mod:
                  	module declares its path as: golang.org/x/example
                  	        but was required as: github.com/golang/example
              stdout:
       
       Summary for local
       -------------
       Succeeded: 16 (changed=15)
       Failed:     2
```

If these could be fixed, I can push CI-based Windows testing on top of this PR, to confirm that everything is working as expected.  Please ping me when it's ready for that.

**Comment by noelmcloughlin** [comment]
@myii Yes I saw that and can replicate the issued on the command-line. 
On the commend line I can replicate that error with go1.16.x but go1.15.9 worked so I updated PR.
Unfortunately I see CI jobs failed with 1.15.9 in gitlab.
I will try reverting to 1.14.2

**Comment by myii** [comment]
@noelmcloughlin OK, I'll see what happens when I run the Windows CI.  In fact, I've noticed that this formula is configured for FreeBSD and OpenBSD as well, so I'll try those at the same time.

**Comment by noelmcloughlin** [comment]
The suite passed after reverting to 1.14.2 so some go later version issue.  This PR should be fine.

**Comment by myii** [comment]
@noelmcloughlin OK, the initial results (some of the GitHub Actions jobs haven't even started yet, though).

[Windows](https://github.com/myii/golang-formula/actions/runs/708224420):

* The `default` suite [fails one state](https://github.com/myii/golang-formula/runs/2245264351?check_suite_focus=true#step:9:1487):

```sls
          ID: golang-package-archive-install-archive-extracted
           Function: archive.extracted
        Name: C:\\golang\\/go1.14.2.windows-amd64
             Result: False
            Comment: Attempt 1: Returned a result of "False", with the following comment: "An exception occurred in this state: Traceback (most recent call last):
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\state.py", line 2172, in call
                  *cdata["args"], **cdata["kwargs"]
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2316, in wrapper
                  return f(*args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\states\archive.py", line 1396, in extracted
                  **kwargs
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
              TypeError: unzip() got an unexpected keyword argument 'uri'
              "
              An exception occurred in this state: Traceback (most recent call last):
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\state.py", line 2172, in call
                  *cdata["args"], **cdata["kwargs"]
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2316, in wrapper
                  return f(*args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\states\archive.py", line 1396, in extracted
                  **kwargs
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 1235, in __call__
                  return self.loader.run(run_func, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2268, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt-3003-py3.7.egg\salt\loader.py", line 2283, in _run_as
                  return _func_or_method(*args, **kwargs)
              TypeError: unzip() got an unexpected keyword argument 'uri'
            Started: 10:44:55.528751
           Duration: 33236.029 ms
            Changes:   
         Name: go clean -i github.com/golang/example/hello... - Function: cmd.run - Result: Clean Started: - 10:45:28.775016 Duration: 24346.473 ms
         Name: go clean -i github.com/golang/example/outyet... - Function: cmd.run - Result: Clean Started: - 10:45:53.121489 Duration: 15.0 ms
         Name: go get github.com/golang/example/hello - Function: cmd.run - Result: Clean Started: - 10:45:53.137488 Duration: 15.0 ms
         Name: go get github.com/golang/example/outyet - Function: cmd.run - Result: Clean Started: - 10:45:53.152488 Duration: 17.011 ms
       
       Summary for local
       ------------
       Succeeded: 7 (changed=3)
       Failed:    1
       ------------
       Total states run:     8
       Total run time:  57.836 s
```

* The `package` suite [only runs the `_mapdata` suite](https://github.com/myii/golang-formula/runs/2245264351?check_suite_focus=true#step:9:2507):

```sls
       Summary for local
       ------------
       Succeeded: 1 (changed=1)
       Failed:    0
       ------------
       Total states run:     1
       Total run time: 180.000 ms
```

Which suites should be running on Windows?

---

I'll respond soon with the *BSD results.

**Comment by myii** [comment]
@noelmcloughlin FreeBSD needed a little fiddling but I've got them working in Vagrant locally for both suites.  I'll push them through soon.  Still need to check OpenBSD and Windows.

**Comment by myii** [comment]
@noelmcloughlin OpenBSD no longer seems to be supported?  No mention at either of these locations:

* https://storage.googleapis.com/golang/
* https://golang.org/dl/

**Comment by myii** [comment]
@noelmcloughlin OpenBSD `package`-based installation works (after a tiny tweak).  So just need to figure out Windows.

File: golang/defaults.yaml [hunk 1]
```diff
@@ -12,15 +12,15 @@ golang:
     name: golang
     use_upstream_repo: false
     use_upstream_archive: true
+    uri: https://storage.googleapis.com/golang
```
Comment (noelmcloughlin) [inline golang/defaults.yaml]: @myii  This was the fix for "_TypeError: unzip() got an unexpected keyword argument 'uri'" ... moving "uri_" up one level.  Not sure why your GitHub Actions is still complaining.

File: golang/defaults.yaml [hunk 2]
```diff
@@ -12,15 +12,15 @@ golang:
...
     archive:
       name: /usr/local
-      uri: https://storage.googleapis.com/golang
       source: null
       source_hash: null
       trim_output: true
-      archive_suffix: tar.gz
       archive_format: tar
       enforce_toplevel: true
+    archive_suffix: tar.gz
```
Comment (noelmcloughlin) [inline golang/defaults.yaml]: This was a fix for "@myii  This was the fix for "TypeError: unzip() got an unexpected keyword argument 'archive_suffix'" too, which I also saw on Windows 10.  I thought these were fixed?

**Comment by myii** [comment]
Well, I've got Windows working with the `package` suite (just need to fix the InSpec tests).  Will have to check the `default` suite afterwards.

File: golang/defaults.yaml (hunk 1 above)
Comment (myii) [reply golang/defaults.yaml]: @noelmcloughlin Well, the `default` (archive) suite is still using `pillar.example` as its pillar:

https://github.com/saltstack-formulas/golang-formula/blob/43b3b932a22dc2bf26e1f23642741d7c6d2931ce/kitchen.yml#L322-L323

And that still has the `uri` defined:

https://github.com/saltstack-formulas/golang-formula/blob/43b3b932a22dc2bf26e1f23642741d7c6d2931ce/pillar.example#L11-L13

I've just tried by commenting that out and the states have passed.  So now need to figure out what the right situation is for _all_ platforms.

Comment (myii) [reply golang/defaults.yaml]: @noelmcloughlin So `uri` and `archive_suffix` only appear to be used for `golang.pkg.use_upstream_archive`:

https://github.com/saltstack-formulas/golang-formula/blob/32e1fc77d4aa351a49f01b335402c665280828f9/golang/map.jinja#L57-L65

So shouldn't all of these occurrences be put underneath that key?

Comment (myii) [reply golang/defaults.yaml]: This is the problem behind everything:

https://github.com/saltstack-formulas/golang-formula/blob/32e1fc77d4aa351a49f01b335402c665280828f9/golang/archive/install.sls#L25-L27

The Windows `archive.extracted` state doesn't handle these values, so this whole section shouldn't be done in this way.  The implementation is messy, so something has to give.  Really, nothing should be under `golang.pkg.archive` except that it appears in the docs under https://docs.saltproject.io/en/latest/ref/states/all/salt.states.archive.html.

Comment (myii) [reply golang/defaults.yaml]: Alright, I've placed those non-state entries under a new `archive_helper` key and that seems to be a lot cleaner.  Testing on all platforms again.

**Comment by myii** [comment]
@noelmcloughlin Please check the commits I added to get FreeBSD, OpenBSD and Windows working in GitHub Actions.  If all is OK and all of the tests pass, feel free to merge this.

---

* Semi-automated using myii/ssf-formula#307

**Review by myii — APPROVED** [review]

File: golang/defaults.yaml (hunk 1 above)
Comment (noelmcloughlin) [reply golang/defaults.yaml]: @myii good idea, I like `archive_helper`.

**Comment by noelmcloughlin** [comment]
Thanks all good.

**Pull request merged**